### PR TITLE
[Fixes #8112] Dump and restore Geofence rules in case of requests or transaction failures

### LINKED ISSUE #8112: Dump and restore Geofence rules in case of requests or transaction failures
## Expected Behavior
When updating/changing the layer permissions rules, it should communicate with Geofence in the following way:
* Purge all Geofence permission rules.
* Create the new rules one by one.
* Restore rules in case of failure.

## Actual Behavior
* Purge all Geofence permission rules.
* Create the new rules one by one.
* If something goes wrong: 
    * The Geofence rules are let in an inconsistent state. For example: the new rules are not created after the purge.

## Steps to Reproduce the Problem
  1. Create a Layer.
  2. Change Layer Permissions.
  3. During the update, add some intentional error:
      * For example: Edit the code.
  4. The rule at Geofence will be in an inconsistent state.

## Specifications

  - GeoNode version: 3.3.x, 3.2.x and also master
  - Installation method (manual, GeoNode Docker, SPCGeoNode Docker): GeoNode Docker
  - Platform:
  - Additional details:

### PR DESCRIPTION
**Pull request by afabiani**
<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [x] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the QA checks: flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

**Comment by italogsfernandes** [comment]
I'll be adding some unit tests here before merge.

**Comment by afabiani** [comment]
@italogsfernandes I migrated the ones from 3.3.x already, waiting for CircleCI to finish

File: geonode/resource/manager.py [hunk 1]
```diff
@@ -558,7 +561,10 @@ def remove_permissions(self, uuid: str, /, instance: ResourceBase = None) -> boo
                 _resource.set_processing_state(enumerations.STATE_PROCESSED)
         return False
 
-    def set_permissions(self, uuid: str, /, instance: ResourceBase = None, owner: settings.AUTH_USER_MODEL = None, permissions: dict = {}, created: bool = False) -> bool:
+    def set_permissions(
+        self, uuid: str, /, instance: ResourceBase = None, owner: settings.AUTH_USER_MODEL = None,
+        permissions: dict = {}, created: bool = False
+    ) -> bool:
```
Comment (italogsfernandes) [inline geonode/resource/manager.py:567]: The argument `geofence_uow: GeofenceLayerRulesUnitOfWork = None` was removed of this method.

Inside the `atomic` block of this method, Geofence is modified by:
* `self.remove_permissions(uuid, instance=_resource)`
* `self._concrete_resource_manager.set_permissions(uuid, instance=_resource, owner=owner, permissions=permissions, created=created)`

Inside each one of then there is still a atomic GeofenceLayerRulesUnitOfWork context block, but without an "outer" geonfence_uow being passed down, the following scenario is still possible:
* Geofence has X rules.
* `remove_permissions` is successfully executed. Now Geofence has 0 rules.
* `self._concrete_resource_manager.set_permissions` raises an exception, so some new rules aren't created. The rollback inside the concrete obj will restore the previous state of 0 rules.
* In this case, the Geofence data won't be restored to X rules as it was before.

Comment (afabiani) [reply geonode/resource/manager.py:567]: @italogsfernandes that's true but in this architecture the generic manager cannot be aware of the specific GeoFence/GeoServer stuff.
We should probably generalize here the concept of security rule and Rule Unit of Work and use the generic class to pass the rules to the concrete manager.
It is a bit out of context for the moment, I prefer to remove the rules if the object is in an inconsistent state and throw an exception instead.

Comment (italogsfernandes) [reply geonode/resource/manager.py:567]: Ok

**Review by italogsfernandes — APPROVED** [review]

**Pull request closed without merging**